packages/paste-website/src/pages/blog/2025-03-20-css-variables-ssr-ssg.mdx

```diff
@@ -0,0 +1,189 @@
+
+#### Fallback or Default Theme During Initial Render
+If you don’t handle the theme properly, Next.js may serve a default theme (e.g., light mode) before applying the correct one. This results in a brief flash where the wrong theme appears before React updates it.
+
+
+### The solution: CSS variables
+
+Using CSS variables with <InlineCode>data-theme</InlineCode> set on the <InlineCode>html</InlineCode> or <InlineCode>body</InlineCode> is the most effective solution as the styles are applied <strong>before</strong> React hydrates the page. This ensures that the correct theme is applied from the start, preventing any flickering or flash.
+
+To support this approach, as of <Anchor showExternal={true} href="https://www.npmjs.com/package/@twilio-paste/design-tokens">@twilio-paste/design-tokens v10.14.0</Anchor> and <Anchor showExternal={true} href="https://www.npmjs.com/package/@twilio-paste/core">@twilio-paste/core v20.23.0</Anchor> we have added a new file to each of our themes called <InlineCode>tokens.data-theme.css</InlineCode> which contains the CSS variables for each theme when the <InlineCode>data-theme</InlineCode> is set.
+
+An example of how this CSS file is structured:
+
+<Box marginBottom="space60">
+  <CodeBlockWrapper>
+    <CodeBlockHeader>
+      @twilio-paste/design-tokens/dist/themes/twilio-dark/tokens.data-theme.css
+    </CodeBlockHeader>
+    <CodeBlock language='javascript' code={`body[data-theme="twilio-dark"] {
+  --color-background-user: rgb(34, 9, 74);
+  --color-background-notification: rgb(214, 31, 31);
+  --color-background-trial: rgb(5, 41, 18);
+  --color-background-subaccount: rgb(18, 28, 45);
+  --color-background-primary-stronger: rgb(204, 228, 255);
+  ...
+    `}/>
+  </CodeBlockWrapper>
+</Box>
+
+We have also added a new prop to the <InlineCode>ThemeProvider</InlineCode> called <InlineCode>useCSSVariables</InlineCode> which will allow you to use CSS variables instead of the theme files we previously configured. Note that we only want to use CSS variables when encoutnering this issue and not as a standard. For all other use cases we recommend using the <InlineCode>theme</InlineCode> prop.
+
+
+## Paste Doc site implementation
+
+This section will cover how we implemented this solution on the Paste documentation site using Next.js.
+
+### _documents.tsx
+
+We added a script to the <InlineCode>head</InlineCode> of our <InlineCode>_documents.tsx</InlineCode> file to set the <InlineCode>data-theme</InlineCode> attribute on the <InlineCode>body</InlineCode> element element using a script. This script will check to see if they have a cookie set and if not check the systme prefernces to determine whether users prefer dark or light themes and apply the dark theme if they do. Otherwise we do not set a value. You will see in <InlineCode>_app.tsx</InlineCode> how we handle the default value.
```

Comment:
```suggestion
We added a script to the <InlineCode>head</InlineCode> of our <InlineCode>_documents.tsx</InlineCode> file to set the <InlineCode>data-theme</InlineCode> attribute on the <InlineCode>body</InlineCode> element using a script. This script will check to see if users have a cookie set and if not check the system preferences to determine whether users prefer dark or light themes, and apply the dark theme if they do. Otherwise we do not set a value. You will see in <InlineCode>_app.tsx</InlineCode> how we handle the default value.
```